src/Cooked/MockChain/Testing.hs

```diff
@@ -295,10 +133,175 @@ assertSubset l r =
+      testPrettyOpts = def
+    }
+
+-- | A test template with no particular requirement on the run
+emptyTest :: (IsProp prop) => StagedMockChain a -> Test a prop
+emptyTest run =
+  Test
+    { testRun = run,
+      testInitDist = def,
+      testErrorProp = \_ _ _ -> testSuccess,
+      testResultProp = \_ _ _ _ -> testSuccess,
+      testPrettyOpts = def
+    }
+
+infixl 5 `withInitDist`
+
+infixl 5 `withPrettyOpts`
+
+infixl 5 `withJournalPred`
+
+infixl 5 `withValueAndStatePred`
+
+infixl 5 `withValuePred`
+
+infixl 5 `withStatePred`
+
+infixl 5 `withPrettyAndErrorPred`
+
+infixl 5 `withErrorPred`
+
+-- | Appending an initial distribution to a test
+withInitDist :: (IsProp prop) => Test a prop -> InitialDistribution -> Test a prop
+withInitDist test initDist = test {testInitDist = initDist}
+
+-- | Appending printing options to a test
+withPrettyOpts :: (IsProp prop) => Test a prop -> PrettyCookedOpts -> Test a prop
+withPrettyOpts test opts = test {testPrettyOpts = opts}
+
+-- | Appending a predicate over the log to a test. This will be used both in
+-- case of success or failure of the run.
+withJournalPred :: (IsProp prop) => Test a prop -> ([MockChainLogEntry] -> prop) -> Test a prop
+withJournalPred test journalPred =
+  test
+    { testErrorProp = \opts err journal -> testErrorProp test opts err journal .&&. journalPred journal,
+      testResultProp = \opts val state journal -> testResultProp test opts val state journal .&&. journalPred journal
+    }
+
+-- | Appending a predicate over the return value and state, which will be used
+-- in case of success of the run.
+withValueAndStatePred :: (IsProp prop) => Test a prop -> (a -> UtxoState -> prop) -> Test a prop
+withValueAndStatePred test resultPred =
+  test
+    { testResultProp = \opts val state journal -> testResultProp test opts val state journal .&&. resultPred val state
+    }
+
+-- | Appending a predicate over the return value, which will be used in case of
+-- success of the run.
+withValuePred :: (IsProp prop) => Test a prop -> (a -> prop) -> Test a prop
+withValuePred test valuePred = withValueAndStatePred test $ \val _ -> valuePred val
+
+-- | Appending a predicate over the return state, which will be used in case of
+-- success of the run.
+withStatePred :: (IsProp prop) => Test a prop -> (UtxoState -> prop) -> Test a prop
+withStatePred test statePred = withValueAndStatePred test $ \_ st -> statePred st
+
+-- | Appending a predicate over an error which uses the printing options, which
+-- will be used in case of failure of the run.
+withPrettyAndErrorPred :: (IsProp prop) => Test a prop -> (PrettyCookedOpts -> MockChainError -> prop) -> Test a prop
+withPrettyAndErrorPred test errorPred = test {testErrorProp = \opts err journal -> testErrorProp test opts err journal .&&. errorPred opts err}
+
+withErrorPred :: (IsProp prop) => Test a prop -> (MockChainError -> prop) -> Test a prop
+withErrorPred test errorPred = withPrettyAndErrorPred test $ \_ err -> errorPred err
```

Comment:
Since we use optics, shouldn't those just be lenses?

I am not fond of involving optics in the user-facing API but conceptually this feels on topic.

Reply:
Ok